src/validator.cpp

```diff
@@ -1249,4 +1259,47 @@ void Validator::ValidatorImpl::checkUnitForCycles(const ModelPtr &model, const U
     }
 }
 
+void Validator::ValidatorImpl::checkUnitHomogeneity(const ModelPtr &model)
+{
+    std::vector<std::pair<VariablePtr, VariablePtr>> checkedPairs;
+    std::vector<std::pair<VariablePtr, VariablePtr>> errorPairs;
+
+    // Check components.
+    if (model->componentCount() > 0) {
+        for (size_t i = 0; i < model->componentCount(); ++i) {
+            auto component = model->component(i);
+            // Check variables.
+            for (size_t j = 0; j < component->variableCount(); ++j) {
+                auto variable = component->variable(j);
+                if (variable->equivalentVariableCount() > 0) {
+                    for (size_t k = 0; k < variable->equivalentVariableCount(); ++k) {
+                        // Check both equivalence pairings
+                        auto equivalentVariable = variable->equivalentVariable(k);
+                        auto firstCheckPairing = std::make_pair(variable, equivalentVariable);
+                        auto secondCheckPairing = std::make_pair(equivalentVariable, variable);
+
+                        if ((std::find(checkedPairs.begin(), checkedPairs.end(), firstCheckPairing) == checkedPairs.end())
+                            && (std::find(checkedPairs.begin(), checkedPairs.end(), secondCheckPairing) == checkedPairs.end())) {
+                            checkedPairs.push_back(firstCheckPairing);
+
+                            // Check equivalence
+                            bool equivalent = variable->units()->equivalent(variable->units(), equivalentVariable->units());
+                            if (!equivalent) {
+                                auto unitsName = variable->units() == nullptr ? "" : variable->units()->name();
+                                auto equivalentUnitsName = equivalentVariable->units() == nullptr ? "" : equivalentVariable->units()->name();
+                                ErrorPtr err = Error::create();
+                                err->setDescription("Error: Variables '" + variable->name() + "' and '" + equivalentVariable->name() + "' do not have the same unit reduction.");
```

Comment:
We don't need the `Error: ` prefix in the error description, one of those that is what it says on the tin things.